# **DS 6001: Lab Assignment 8**
## **Randa Ampah**

In [3]:
import numpy as np
import pandas as pd
import sidetable
import sqlite3

## **Problem 1**

### **Part A**

I like the way that polars selects data within dataframes, only because I sometimes have trouble remembering the syntax for slicing in pandas. I also like that this structure is similar to R (specifically tidyverse), because I sometimes forget what function is used in R vs Python for functions like this.

I find the column assignment based on predicate to be interesting. Looking objectively as an outsider to both chunks of code, polars is definitely more intuitive to understand in terms of the conditional assignment structure.



### **Part B**

The best argument for pandas is that it is more established and better suited for higher level machine learning integration, as well as exploration of data, which makes is more relevant as of now.

### **Part C**

Part of a comment from a deleted user on Reddit stated "So that's what I would base my choice on. If you're interested in how big data applications handle data then I would go with Polars. If you're just interested in the practical aspect of getting something working and you want lots of resources and examples to help you use the tool, then Pandas is probably the better choice."

From what I have gathered about polars from the previous readings, I would say that I agree with this summary of the main differences between polars and pandas. A lot of the merits highlighted in the polars user guide were focused more on computation speed and efficiency, which is undoubtedly important, but is something more likely to be honed by a more established data scientist. As a beginner data scientist getting my bearings in the analytical approaches and tools utilized in the field, I would want to focus on pandas for now.

### **Part D**

Speaking to two of my classmates, they both were a bit open to learning polars. One said that she doesn't mind learning polars because it would be another tool in her arsenal, while the other said she prefers pandas for now since she isn't familiar with polars yet, but she is open to utilizing it down the line.

### **Part E**

Through exploring the merits and deficits of polars, I am intrigued by polars and I would consider learning to use it down the line, when I more regularly work with larger datasets. However, for now, I would like to focus on strengthening my pandas skills and learning to utilize it in a more efficient way, as well as with a more analytical mind.

## **Problem 2**

### **Part A**

In [5]:
url = 'https://www.cdc.gov/workplace-health-promotion/media/files/2024/06/whpps_120717.csv'
df = pd.read_csv(url, delimiter='~')
df.head()

/var/folders/1z/t_vl0dx95552hzn7xv8z08sm0000gn/T/ipykernel_1482/3374536002.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(url, delimiter='~')


OC1  OC3  HI1  HI2  HI3  HI4  HRA1  HRA1A  HRA1B  HRA1E  ...  WL3_05  E1_09  \
0   7  3.0  2.0  1.0  2.0  1.0   1.0    3.0    4.0    2.0  ...     PTO    NaN   
1   2  3.0  2.0  3.0  1.0  1.0   1.0    3.0    3.0    1.0  ...     NaN    NaN   
2   7  3.0  1.0  3.0  1.0  1.0   1.0    3.0   97.0    2.0  ...     NaN    NaN   
3   1  2.0  1.0  2.0  1.0  1.0  97.0   96.0   96.0   96.0  ...     NaN    NaN   
4   2  3.0  1.0  3.0  1.0  1.0   1.0    3.0    3.0    2.0  ...     NaN    NaN   

   Suppquex     Id  Region  CDC_Region  Industry  Size  Varstrata  \
0       2.0  217.0     1.0         2.0       7.0   7.0        0.0   
1       1.0  326.0     3.0         7.0       7.0   6.0        0.0   
2       1.0  399.0     4.0         8.0       7.0   8.0        0.0   
3       1.0  475.0     5.0         9.0       7.0   4.0        0.0   
4       1.0  489.0     2.0         4.0       7.0   4.0        0.0   

   Finalwt_worksite,,,,  
0      47.793940929,,,,  
1      47.793940929,,,,  
2      47.793940929,,,,  
3      47.793940929,,,,  
4      47.793940929,,,,  

[5 rows x 301 columns]

In [40]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2843 entries, 0 to 2842
Data columns (total 301 columns):
 #    Column                Dtype  
---   ------                -----  
 0    OC1                   object 
 1    OC3                   float64
 2    HI1                   float64
 3    HI2                   float64
 4    HI3                   float64
 5    HI4                   float64
 6    HRA1                  float64
 7    HRA1A                 float64
 8    HRA1B                 float64
 9    HRA1E                 float64
 10   CP1                   float64
 11   CP2                   float64
 12   CP3                   float64
 13   CP4                   float64
 14   CP5                   float64
 15   HP1                   float64
 16   HP2                   float64
 17   HP3                   float64
 18   HP4                   float64
 19   HP5                   float64
 20   HP5A                  float64
 21   HP6                   float64
 22   HP7A                  

The data has 2,843 rows and 301 columns. There are 294 float columns and 7 object columns.

### **Part B**

In [38]:
df.columns[df.columns.str.contains('WD')]

Index(['WD1_1', 'WD1_2', 'WD2', 'WD3', 'WD4', 'WD5', 'WD6', 'WD7'], dtype='object')

In [11]:
whpps = df[['Industry','Size','OC3','HI1','HI2','HI3','CP1',
            'WL6','WD1_1', 'WD1_2', 'WD2', 'WD3', 'WD4', 'WD5',
            'WD6', 'WD7']]
whpps.head()

Industry  Size  OC3  HI1  HI2  HI3  CP1  WL6  WD1_1  WD1_2    WD2    WD3  \
0       7.0   7.0  3.0  2.0  1.0  2.0  1.0  1.0   25.0   20.0   85.0   60.0   
1       7.0   6.0  3.0  2.0  3.0  1.0  1.0  1.0  997.0  997.0   90.0   90.0   
2       7.0   8.0  3.0  1.0  3.0  1.0  1.0  1.0   35.0    4.0  997.0  997.0   
3       7.0   4.0  2.0  1.0  2.0  1.0  2.0  2.0   50.0   15.0   50.0   85.0   
4       7.0   4.0  3.0  1.0  3.0  1.0  1.0  1.0   50.0   40.0   60.0   60.0   

     WD4    WD5    WD6    WD7  
0   40.0   15.0    0.0   22.0  
1  997.0  997.0    0.0  997.0  
2   40.0   15.0  997.0  997.0  
3   75.0    0.0    0.0  997.0  
4   40.0   30.0    0.0   28.0

### **Part C**

In [12]:
whpps = whpps.rename({'OC3': 'OrgStructure',
                      'HI1': 'InsFTEmp',
                      'HI2': 'InsFTPremiumChange',
                      'HI3': 'InsPTEmp',
                      'CP1': 'HealthEdOffered',
                      'WL6': 'WFHAllowed',
                      'WD1_1': 'PctEmpUnder30', 
                      'WD1_2': 'PctEmpOver60', 
                      'WD2': 'PctEmpFemale', 
                      'WD3': 'PctEmpHourly', 
                      'WD4': 'PctEmpNotDayShift', 
                      'WD5': 'PctEmpRemoteOffSite', 
                      'WD6': 'PctEmpUnion', 
                      'WD7': 'AvgYrTurnoverPct'}, axis = 1)

whpps.head()

Industry  Size  OrgStructure  InsFTEmp  InsFTPremiumChange  InsPTEmp  \
0       7.0   7.0           3.0       2.0                 1.0       2.0   
1       7.0   6.0           3.0       2.0                 3.0       1.0   
2       7.0   8.0           3.0       1.0                 3.0       1.0   
3       7.0   4.0           2.0       1.0                 2.0       1.0   
4       7.0   4.0           3.0       1.0                 3.0       1.0   

   HealthEdOffered  WFHAllowed  PctEmpUnder30  PctEmpOver60  PctEmpFemale  \
0              1.0         1.0           25.0          20.0          85.0   
1              1.0         1.0          997.0         997.0          90.0   
2              1.0         1.0           35.0           4.0         997.0   
3              2.0         2.0           50.0          15.0          50.0   
4              1.0         1.0           50.0          40.0          60.0   

   PctEmpHourly  PctEmpNotDayShift  PctEmpRemoteOffSite  PctEmpUnion  \
0          60.0               40.0                 15.0          0.0   
1          90.0              997.0                997.0          0.0   
2         997.0               40.0                 15.0        997.0   
3          85.0               75.0                  0.0          0.0   
4          60.0               40.0                 30.0          0.0   

   AvgYrTurnoverPct  
0              22.0  
1             997.0  
2             997.0  
3             997.0  
4              28.0

### **Part D**

#### Step I

In [13]:
replace_map = {
    1:'Food, Power, Utility Acquisition and Construction',
    2:'Trade and Logistics',
    3:'Leisure and Other',
    4:'Corporate Services',
    5:'Healthcare and Education',
    6:'Public Admin',
    7:'Hospital'
}

whpps['Industry'] = whpps['Industry'].replace(replace_map)
whpps['Industry'].value_counts()

Industry
Healthcare and Education                             551
Food, Power, Utility Acquisition and Construction    525
Leisure and Other                                    433
Corporate Services                                   429
Hospital                                             338
Trade and Logistics                                  311
Public Admin                                         255
Name: count, dtype: int64

#### Step II

In [14]:
size_replace = {1: 'Small',
                2: 'Small',
                3: 'Small',
                4: 'Medium',
                5: 'Medium',
                6: 'Large',
                7: 'Large',
                8: 'Large'}

whpps['Size'] = whpps['Size'].replace(size_replace)
whpps['Size'].value_counts(dropna=False)

Size
Small     2195
Medium     393
Large      254
NaN          1
Name: count, dtype: int64

#### Step III

In [15]:
whpps.dtypes

Industry                object
Size                    object
OrgStructure           float64
InsFTEmp               float64
InsFTPremiumChange     float64
InsPTEmp               float64
HealthEdOffered        float64
WFHAllowed             float64
PctEmpUnder30          float64
PctEmpOver60           float64
PctEmpFemale           float64
PctEmpHourly           float64
PctEmpNotDayShift      float64
PctEmpRemoteOffSite    float64
PctEmpUnion            float64
AvgYrTurnoverPct       float64
dtype: object

In [16]:
replace_map = {'OrgStructure':{1:'For Profit, Public',
                               2:'For Profit, Private',
                               3:'Non-Profit',
                               4:'State or Local Government',
                               5:'Federal Government',
                               6:'Other',
                               97: np.nan,
                               98: np.nan,
                               99: np.nan},
                'InsFTEmp':  {1:'Full Insurance Offered',
                              2:'Partial Insurance Offered',
                              3:'No Insurance Offered',
                              97: np.nan,
                              98: np.nan,
                              99: np.nan},
                'InsFTPremiumChange': {1:'Larger',
                                       2:'Smaller',
                                       3:'About the Same',
                                       96: np.nan,
                                       97: np.nan,
                                       98: np.nan,
                                       99: np.nan},
                'InsPTEmp': {1:'Yes',
                             2:'No',
                             97: np.nan,
                             98: np.nan,
                             99: np.nan},
                'HealthEdOffered': {1:'Yes',
                                    2:'No',
                                    97: np.nan,
                                    98: np.nan},
                'WFHAllowed': {1:'Yes',
                               2:'No',
                               97: np.nan,
                               98: np.nan,
                               99: np.nan}
                }

whpps = whpps.replace(replace_map)

#### Step IV

In [17]:
whpps.dtypes

Industry                object
Size                    object
OrgStructure            object
InsFTEmp                object
InsFTPremiumChange      object
InsPTEmp                object
HealthEdOffered         object
WFHAllowed              object
PctEmpUnder30          float64
PctEmpOver60           float64
PctEmpFemale           float64
PctEmpHourly           float64
PctEmpNotDayShift      float64
PctEmpRemoteOffSite    float64
PctEmpUnion            float64
AvgYrTurnoverPct       float64
dtype: object

In [18]:
whpps.select_dtypes(include='float64').columns

Index(['PctEmpUnder30', 'PctEmpOver60', 'PctEmpFemale', 'PctEmpHourly',
       'PctEmpNotDayShift', 'PctEmpRemoteOffSite', 'PctEmpUnion',
       'AvgYrTurnoverPct'],
      dtype='object')

In [19]:
num_cols = whpps.select_dtypes(include='float64').columns

whpps[num_cols] = whpps[num_cols].replace([997,998,999],np.nan)

### **Part E**

In [20]:
whpps = whpps.sort_values(by=['Industry','Size','PctEmpUnder30'],
                  ascending=[True,True,False])
# truncated for sake of submission
whpps.head(10)

Industry    Size         OrgStructure  \
465   Corporate Services   Large   For Profit, Public   
2471  Corporate Services   Large  For Profit, Private   
1352  Corporate Services   Large  For Profit, Private   
821   Corporate Services   Large  For Profit, Private   
822   Corporate Services   Large           Non-Profit   
1872  Corporate Services   Large                  NaN   
1585  Corporate Services   Large                  NaN   
1351  Corporate Services   Large  For Profit, Private   
2468  Corporate Services  Medium  For Profit, Private   
2229  Corporate Services  Medium  For Profit, Private   

                       InsFTEmp InsFTPremiumChange InsPTEmp HealthEdOffered  \
465      Full Insurance Offered     About the Same       No             Yes   
2471  Partial Insurance Offered             Larger       No             Yes   
1352     Full Insurance Offered     About the Same       No             Yes   
821   Partial Insurance Offered     About the Same       No             Yes   
822      Full Insurance Offered     About the Same       No             Yes   
1872  Partial Insurance Offered             Larger       No              No   
1585  Partial Insurance Offered             Larger      Yes             Yes   
1351  Partial Insurance Offered     About the Same       No             Yes   
2468  Partial Insurance Offered     About the Same       No             Yes   
2229  Partial Insurance Offered     About the Same      Yes              No   

     WFHAllowed  PctEmpUnder30  PctEmpOver60  PctEmpFemale  PctEmpHourly  \
465         Yes           75.0           1.0          49.0          25.0   
2471        Yes           70.0          30.0           NaN           NaN   
1352        Yes           40.0           5.0          30.0          30.0   
821         Yes           35.0          15.0          45.0          75.0   
822          No           34.0          10.0          86.0          88.0   
1872         No           30.0          20.0          78.0          75.0   
1585        Yes           15.0          25.0          25.0          40.0   
1351        Yes           11.0           8.0          58.0          11.0   
2468         No           80.0           1.0          10.0          97.0   
2229        Yes           50.0           1.0          28.0           7.0   

      PctEmpNotDayShift  PctEmpRemoteOffSite  PctEmpUnion  AvgYrTurnoverPct  
465                25.0                 18.0          0.0              11.0  
2471               20.0                  3.0          0.0               NaN  
1352               30.0                 25.0          0.0               NaN  
821                40.0                 45.0         70.0              10.0  
822                 NaN                  1.0          1.0               NaN  
1872                1.0                  4.0          0.0              35.0  
1585               10.0                 10.0         20.0              10.0  
1351                NaN                  NaN          0.0               NaN  
2468                1.0                  0.0          0.0               8.0  
2229                3.0                 60.0          0.0              15.0

### **Part F**

In [21]:
whpps = whpps.query('Industry.isna() == False')
len(whpps)

2842

### **Part G**

In [22]:
whpps['GenderBalance'] = pd.cut(whpps.PctEmpFemale, 
                                bins=[-0.1,35,65,100],
                                labels = ('Mostly men','Balanced','Mostly women'))

whpps['GenderBalance'].value_counts()

GenderBalance
Mostly men      733
Mostly women    727
Balanced        587
Name: count, dtype: int64

### **Part H**

In [23]:
whpps.dtypes

Industry                 object
Size                     object
OrgStructure             object
InsFTEmp                 object
InsFTPremiumChange       object
InsPTEmp                 object
HealthEdOffered          object
WFHAllowed               object
PctEmpUnder30           float64
PctEmpOver60            float64
PctEmpFemale            float64
PctEmpHourly            float64
PctEmpNotDayShift       float64
PctEmpRemoteOffSite     float64
PctEmpUnion             float64
AvgYrTurnoverPct        float64
GenderBalance          category
dtype: object

In [24]:
whpps.columns

Index(['Industry', 'Size', 'OrgStructure', 'InsFTEmp', 'InsFTPremiumChange',
       'InsPTEmp', 'HealthEdOffered', 'WFHAllowed', 'PctEmpUnder30',
       'PctEmpOver60', 'PctEmpFemale', 'PctEmpHourly', 'PctEmpNotDayShift',
       'PctEmpRemoteOffSite', 'PctEmpUnion', 'AvgYrTurnoverPct',
       'GenderBalance'],
      dtype='object')

In [25]:
cat_vars = whpps.select_dtypes(include='object').columns
cat_vars

Index(['Industry', 'Size', 'OrgStructure', 'InsFTEmp', 'InsFTPremiumChange',
       'InsPTEmp', 'HealthEdOffered', 'WFHAllowed'],
      dtype='object')

In [26]:
whpps[cat_vars] = whpps[cat_vars].astype('category')

In [27]:
whpps.dtypes

Industry               category
Size                   category
OrgStructure           category
InsFTEmp               category
InsFTPremiumChange     category
InsPTEmp               category
HealthEdOffered        category
WFHAllowed             category
PctEmpUnder30           float64
PctEmpOver60            float64
PctEmpFemale            float64
PctEmpHourly            float64
PctEmpNotDayShift       float64
PctEmpRemoteOffSite     float64
PctEmpUnion             float64
AvgYrTurnoverPct        float64
GenderBalance          category
dtype: object

In [28]:
whpps.head()

Industry   Size         OrgStructure  \
465   Corporate Services  Large   For Profit, Public   
2471  Corporate Services  Large  For Profit, Private   
1352  Corporate Services  Large  For Profit, Private   
821   Corporate Services  Large  For Profit, Private   
822   Corporate Services  Large           Non-Profit   

                       InsFTEmp InsFTPremiumChange InsPTEmp HealthEdOffered  \
465      Full Insurance Offered     About the Same       No             Yes   
2471  Partial Insurance Offered             Larger       No             Yes   
1352     Full Insurance Offered     About the Same       No             Yes   
821   Partial Insurance Offered     About the Same       No             Yes   
822      Full Insurance Offered     About the Same       No             Yes   

     WFHAllowed  PctEmpUnder30  PctEmpOver60  PctEmpFemale  PctEmpHourly  \
465         Yes           75.0           1.0          49.0          25.0   
2471        Yes           70.0          30.0           NaN           NaN   
1352        Yes           40.0           5.0          30.0          30.0   
821         Yes           35.0          15.0          45.0          75.0   
822          No           34.0          10.0          86.0          88.0   

      PctEmpNotDayShift  PctEmpRemoteOffSite  PctEmpUnion  AvgYrTurnoverPct  \
465                25.0                 18.0          0.0              11.0   
2471               20.0                  3.0          0.0               NaN   
1352               30.0                 25.0          0.0               NaN   
821                40.0                 45.0         70.0              10.0   
822                 NaN                  1.0          1.0               NaN   

     GenderBalance  
465       Balanced  
2471           NaN  
1352    Mostly men  
821       Balanced  
822   Mostly women

## **Problem 3**

### **Part A**

In [29]:
import warnings
warnings.filterwarnings('ignore')

In [30]:
small_wfh = whpps.query("Size == 'Small' and WFHAllowed == 'Yes'")
small_wfh

Industry   Size         OrgStructure  \
2444   Corporate Services  Small  For Profit, Private   
2448   Corporate Services  Small                  NaN   
805    Corporate Services  Small  For Profit, Private   
2688   Corporate Services  Small  For Profit, Private   
2214   Corporate Services  Small                  NaN   
...                   ...    ...                  ...   
1768  Trade and Logistics  Small                Other   
2109  Trade and Logistics  Small                  NaN   
2112  Trade and Logistics  Small                Other   
2384  Trade and Logistics  Small                  NaN   
2629  Trade and Logistics  Small   For Profit, Public   

                       InsFTEmp InsFTPremiumChange InsPTEmp HealthEdOffered  \
2444  Partial Insurance Offered     About the Same       No             Yes   
2448     Full Insurance Offered     About the Same      Yes             NaN   
805   Partial Insurance Offered     About the Same       No              No   
2688     Full Insurance Offered     About the Same       No              No   
2214  Partial Insurance Offered            Smaller       No              No   
...                         ...                ...      ...             ...   
1768  Partial Insurance Offered                NaN      Yes             NaN   
2109  Partial Insurance Offered            Smaller       No              No   
2112     Full Insurance Offered     About the Same       No              No   
2384                        NaN                NaN      Yes             Yes   
2629     Full Insurance Offered             Larger       No             Yes   

     WFHAllowed  PctEmpUnder30  PctEmpOver60  PctEmpFemale  PctEmpHourly  \
2444        Yes           90.0           0.0          50.0           0.0   
2448        Yes           90.0           1.0          40.0          95.0   
805         Yes           80.0           0.0          99.0          90.0   
2688        Yes           80.0           0.0          20.0           0.0   
2214        Yes           75.0           4.0          16.0          28.0   
...         ...            ...           ...           ...           ...   
1768        Yes            NaN           NaN          15.0          80.0   
2109        Yes            NaN           NaN           NaN           NaN   
2112        Yes            NaN           NaN           NaN           NaN   
2384        Yes            NaN           NaN           NaN          20.0   
2629        Yes            NaN           2.0          15.0           NaN   

      PctEmpNotDayShift  PctEmpRemoteOffSite  PctEmpUnion  AvgYrTurnoverPct  \
2444                0.0                  1.0          0.0               NaN   
2448                0.0                  0.0          NaN               5.0   
805                10.0                 10.0          0.0               2.0   
2688                5.0                  0.0          0.0               2.0   
2214                NaN                  0.0          0.0               NaN   
...                 ...                  ...          ...               ...   
1768                NaN                  NaN          NaN               NaN   
2109                NaN                  NaN          NaN               NaN   
2112                0.0                  5.0          NaN               NaN   
2384                NaN                  NaN          NaN               1.0   
2629                NaN                 90.0          0.0              15.0   

     GenderBalance  
2444      Balanced  
2448      Balanced  
805   Mostly women  
2688    Mostly men  
2214    Mostly men  
...            ...  
1768    Mostly men  
2109           NaN  
2112           NaN  
2384           NaN  
2629    Mostly men  

[709 rows x 17 columns]

In [31]:
small_wfh.stb.freq(['InsFTEmp'])

InsFTEmp  count    percent  cumulative_count  \
0     Full Insurance Offered    324  46.285714               324   
1  Partial Insurance Offered    310  44.285714               634   
2       No Insurance Offered     66   9.428571               700   

   cumulative_percent  
0           46.285714  
1           90.571429  
2          100.000000

### **Part B**

In [32]:
whppsdb = sqlite3.connect('whpps,db')

In [33]:
whpps.to_sql('whpps', whppsdb, index=False, chunksize=1000, if_exists='replace')

2842

In [34]:
# Check that database was created properly

bq = '''
SELECT *
FROM whpps
'''
pd.read_sql_query(bq,con=whppsdb).head()

Industry   Size         OrgStructure                   InsFTEmp  \
0  Corporate Services  Large   For Profit, Public     Full Insurance Offered   
1  Corporate Services  Large  For Profit, Private  Partial Insurance Offered   
2  Corporate Services  Large  For Profit, Private     Full Insurance Offered   
3  Corporate Services  Large  For Profit, Private  Partial Insurance Offered   
4  Corporate Services  Large           Non-Profit     Full Insurance Offered   

  InsFTPremiumChange InsPTEmp HealthEdOffered WFHAllowed  PctEmpUnder30  \
0     About the Same       No             Yes        Yes           75.0   
1             Larger       No             Yes        Yes           70.0   
2     About the Same       No             Yes        Yes           40.0   
3     About the Same       No             Yes        Yes           35.0   
4     About the Same       No             Yes         No           34.0   

   PctEmpOver60  PctEmpFemale  PctEmpHourly  PctEmpNotDayShift  \
0           1.0          49.0          25.0               25.0   
1          30.0           NaN           NaN               20.0   
2           5.0          30.0          30.0               30.0   
3          15.0          45.0          75.0               40.0   
4          10.0          86.0          88.0                NaN   

   PctEmpRemoteOffSite  PctEmpUnion  AvgYrTurnoverPct GenderBalance  
0                 18.0          0.0              11.0      Balanced  
1                  3.0          0.0               NaN          None  
2                 25.0          0.0               NaN    Mostly men  
3                 45.0         70.0              10.0      Balanced  
4                  1.0          1.0               NaN  Mostly women

### **Part C**

In [35]:
cq = whpps.query("Industry == 'Hospital' and InsFTPremiumChange == 'Smaller'")
cq_df = cq[['Size','OrgStructure','InsFTEmp','InsPTEmp']]
cq_df = cq_df.rename({'InsFTEmp':'Insurance'}, axis=1)
cq_df

Size         OrgStructure                  Insurance InsPTEmp
187   Large           Non-Profit  Partial Insurance Offered      Yes
97    Large           Non-Profit  Partial Insurance Offered      Yes
214   Large           Non-Profit  Partial Insurance Offered      Yes
3    Medium  For Profit, Private     Full Insurance Offered      Yes
75   Medium           Non-Profit     Full Insurance Offered      NaN
320  Medium           Non-Profit     Full Insurance Offered      Yes
11   Medium                  NaN  Partial Insurance Offered      Yes
48   Medium           Non-Profit  Partial Insurance Offered      Yes
51   Medium           Non-Profit     Full Insurance Offered      Yes
191  Medium           Non-Profit  Partial Insurance Offered      Yes
229   Small           Non-Profit     Full Insurance Offered      Yes

### **Part D**

In [36]:
whpps.groupby('Industry').agg({'PctEmpFemale':'mean',
                               'PctEmpUnder30':'mean',
                               'PctEmpOver60':'mean'}
                               ).reset_index().sort_values(by='PctEmpFemale', ascending=False)

Industry  PctEmpFemale  \
2                           Healthcare and Education     80.657143   
3                                           Hospital     76.427027   
4                                  Leisure and Other     53.804416   
0                                 Corporate Services     50.632184   
5                                       Public Admin     39.056738   
6                                Trade and Logistics     32.657258   
1  Food, Power, Utility Acquisition and Construction     20.328605   

   PctEmpUnder30  PctEmpOver60  
2      25.745665     11.349570  
3      27.213793     16.489655  
4      38.566343     11.544872  
0      23.821752     12.465465  
5      21.015625     15.015385  
6      29.108696     12.584034  
1      22.257143     10.690355

### **Part E**

In [37]:
eq = whpps.query("GenderBalance.isna() == False and InsFTEmp.isna() == False")
eq = eq.groupby(['GenderBalance','InsFTEmp']).size().reset_index()
eq = eq.rename({0:'Count'}, axis=1)
eq

GenderBalance                   InsFTEmp  Count
0    Mostly men     Full Insurance Offered    301
1    Mostly men       No Insurance Offered     91
2    Mostly men  Partial Insurance Offered    332
3      Balanced     Full Insurance Offered    226
4      Balanced       No Insurance Offered     77
5      Balanced  Partial Insurance Offered    271
6  Mostly women     Full Insurance Offered    267
7  Mostly women       No Insurance Offered    107
8  Mostly women  Partial Insurance Offered    333